Add Xiaomi Aqara Config Flow

starkillerOG opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
YAML configuration of the xiaomi_aqara component has been deprecated, please remove the xiaomi_aqara platform from your configuration.yaml file.

The xiaomi_aqara component now uses the new Config Flow.
Xiaomi Aqara Gateways schould be discovered automatically and show up under "Configuration"-> "Integrations", from there hit configure and go through the steps.
If your Xiaomi Aqara Gateway does not show up automatically, go to "Configuration"-> "Integrations"-> press the "+" icon -> search for "xiaomi_aqara" and enter the setup.

If no key is provided during setup, only the binary_sensor and sensor platforms will be available. A key is required to activate the other platforms.

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

- Implement Config Flow
- Implement the zeroconf discovery using Config Flow
- Set device_ids and add device information
- Removed the depricated discovery implementation.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

Config Flow :)

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13453

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

starkillerOG committed:
Xiaomi Aqara Config Flow

starkillerOG committed:
Xiaomi Aqara Config Flow

starkillerOG committed:
Xiaomi Aqara Config Flow

starkillerOG committed:
Xiaomi Aqara Config Flow

starkillerOG committed:
Xiaomi Aqara Config Flow

First tested and working version

starkillerOG committed:
Remove depricated discovery

starkillerOG commented:
To Do:

- add config flow tests
- add documentation
- update .coveragerc
- add async_unload_entry

starkillerOG commented:
@Danielhiversen I am not sure if the XiaomiGateway class of the pyxiaomigateway library does I/O?
If so, how can I schedule the initialization of the class in the executor pool?
The normal `hass.async_add_executor_job` will give me an error that XiaomiGateway is not callable since it is a class and not a function...

MartinHjelmare commented:
```py
gateway_instance = await hass.async_add_executor_job(
    gateway_class, param1, param2
)
```

starkillerOG committed:
Add Xiaomi Aqara Config Flow

starkillerOG commented on homeassistant/components/xiaomi_aqara/__init__.py:
```diff
@@ -15,40 +15,38 @@
 from homeassistant.helpers.event import async_track_point_in_utc_time
 from homeassistant.util.dt import utcnow
 
+from .config_flow import (
```
done, thanks

starkillerOG committed:
Add Xiaomi Aqara tests

starkillerOG committed:
Update .coveragerc

starkillerOG committed:
Update requirements_test_all.txt

starkillerOG committed:
fix spelling mistake

starkillerOG committed:
fix select scheme

starkillerOG commented:
@MartinHjelmare this lint test failure is totally unrelated:
`tests/components/arcam_fmj/test_device_trigger.py:52:26: F541 f-string is missing placeholders`
Think this is something wrong in the dev branch or something.

Could you restart the tests?

MartinHjelmare commented:
Please rebase on latest dev branch to let the build pass.

starkillerOG commented:
@MartinHjelmare This is ready to be reviewed/merged now

starkillerOG committed:
Merge branch 'dev' into patch-26

starkillerOG committed:
fix wrong conflict resolve

starkillerOG committed:
add IP to zeroconf discovery title

starkillerOG committed:
Merge branch 'patch-26' of https://github.com/starkillerOG/home-assistant into patch-26

starkillerOG commented:
@bdraco would you be willing to revieuw this?

starkillerOG committed:
black styling

frenck commented:
@starkillerOG Don't tag random people to ask for reviews. That is not appreciated as all. The PR load is high, please have patience. Thanks 👍

bieniu commented on homeassistant/components/xiaomi_aqara/manifest.json:
```diff
@@ -1,8 +1,10 @@
   "name": "Xiaomi Gateway (Aqara)",
+  "config_flow": true,
   "documentation": "https://www.home-assistant.io/integrations/xiaomi_aqara",
   "requirements": ["PyXiaomiGateway==0.12.4"],
```
We use `getmac` so:
```suggestion
  "requirements": ["PyXiaomiGateway==0.12.4", "getmac==0.8.2"],
```

starkillerOG committed:
add getmac requirement

Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>

starkillerOG committed:
add getmac

starkillerOG committed:
add getmac

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/config_flow.py:
```diff
@@ -0,0 +1,188 @@
+            self.selected_gateway = self.gateways[ip_adress]
+            return await self.async_step_settings()
+
+        select_scheme = vol.Schema(
```
```suggestion
        select_schema = vol.Schema(
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/config_flow.py:
```diff
@@ -0,0 +1,188 @@
+        )
+
+        return self.async_show_form(
+            step_id="select", data_schema=select_scheme, errors=errors
```
```suggestion
            step_id="select", data_schema=select_schema, errors=errors
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/config_flow.py:
```diff
@@ -0,0 +1,188 @@
+        if not name or not self.host or not mac_address:
+            return self.async_abort(reason="not_xiaomi_aqara")
+
+        # format mac (include semicolns and make uppercase)
+        mac_address = format_mac(mac_address)
```
Please move this below the guard below.

starkillerOG replied:
done

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/config_flow.py:
```diff
@@ -0,0 +1,188 @@
+GATEWAY_SETTINGS = vol.Schema(
+    {
+        vol.Optional(CONF_KEY): vol.All(str, vol.Length(min=16, max=16)),
+        vol.Optional(CONF_DISCOVERY_RETRY, default=DEFAULT_DISCOVERY_RETRY): vol.All(
```
Discovery retry should be part of an options flow and not the main flow.

starkillerOG replied:
Schould we not just remove that options and set a sensible default (I think 3 times is enough)?
I don't feel this is actually really nessesary.

If you want I will make a options flow now I know how to do this, but I don't feel the need for it.

MartinHjelmare replied:
Yeah, if we don't need the option please remove it. That's even better.

starkillerOG replied:
I have removed that option and set the defaut to 5 retries

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/config_flow.py:
```diff
@@ -0,0 +1,188 @@
+            protocol = self.selected_gateway.proto
+
+            # get mac address
+            mac_address = await self.async_get_mac(ip_adress)
```
Why can't we use the sid as unique id for the config entry? We don't want to use get_mac to look up mac address since that's not reliable.

starkillerOG replied:
I think because the zeroconf discovery does not provide a sid and only a mac, but I will check this tonight.
Forgot the exact reason why I did it like this.....
If sid can be retrieved from zeroconf discovery I will use that

MartinHjelmare replied:
Maybe we can query the gateway discovery in the zeroconf step and get the sid that way?

starkillerOG replied:
I checked it and indeed the zeroconf discovery_info includes a mac_adress but does not include the sid....

starkillerOG replied:
There is no direct way to get the sid from the gateway even if you already know the host/ip.
To get that you need to do a custom discovery of the gateway using the xiaomi_gateway library. Basically you need to sent UDP message `{"cmd":"whois"}` at multicast adresss '224.0.0.50' gateway_discovery_port 4321. Then the library wait for 5 seconds to gether all responses. Because you know the host already from the zeroconf discovery you can then match one of the gateways that responded to the zeroconf discovered gateway and then you know the required sid.

However that would include that 5 second discovery process inside the async_step_zeroconf, and I did not like that because if people do not use the xiaomi_aqara integration or have already configured it, I think it will go through that whole 5 second discovery process each time a zeroconf scan is done to then find out the SID is already known and ignored or configured. I think that might be a waste of resources.

That is why I opted to use the mac_adress for the unique_id because the mac is included in the zeroconf discovery_info, so no additional device comunication or discovery is required to then abort the config flow if the device is already set up or ignored using the unique_id based on mac.

starkillerOG replied:
@MartinHjelmare If you like me to implement that 5 second discovery process inside the async_step_zeroconf to use sid as unique_id I will.
You can decide based on my previous comment....

starkillerOG replied:
This is the zeroconf discovery_info:
`{'host': '192.168.1.IP', 'port': 54321, 'hostname': 'lumi-gateway-v3_miio12345678.local.', 'type': '_miio._udp.local.', 'name': 'lumi-gateway-v3_miio12345678._miio._udp.local.', 'properties': {'_raw': {'epoch': b'1', 'mac': b'abcdefghifjk'}, 'epoch': '1', 'mac': 'abcdefghifjk'}}`

MartinHjelmare replied:
Since we can't get a unique_id easily, I think we should remove the zeroconf step here and let the default zeroconf step handle the zeroconf discovery. That will allow zeroconf discovery until a config entry is created for this domain.

We can then change to use the `sid` as unique_id for the user configured config entries.

starkillerOG replied:
The default zeroconf step will not work since that will detect lots of other xiaomi products that are not Xiaomi Gateways and can therefore not be used with this integration.
Other products that will be discovered on the same zeroconf service but be filtered out:
- xiaomi air purifiers
- xioami air humidifiers
- xiaomi alarm clock
- xiaomi camera's
- xiaomi speakers
- xiaomi robot vacuum cleaners
- xioami wifi plugs/switches/sockets
and lots more....

This integration only handels xiaomi gateways and there connected zigbee devices.

starkillerOG replied:
@MartinHjelmare I went though everything again and I discovered that I am stupid.
It turns out the sid and the mac are the same thing....
Xiaomi uses sid as device identification, but for the gateway they apperently use the mac_adress as identification. so for a gateway the sid and mac are the same.

Other devices such as the connected zigbee devices do not have an mac, but do have a sid which also has a larger number of characters than a mac.

Anyway I will update the code since sid and mac can be used interchangable for the gateway....

starkillerOG replied:
I removed the getmac depandance

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/cover.py:
```diff
@@ -13,29 +14,31 @@
-    add_entities(devices)
+    entities = []
+    gateway = hass.data[DOMAIN][config_entry.entry_id]
+    for entity in gateway.devices["cover"]:
+        model = entity["model"]
+        if model in ["curtain", "curtain.aq2", "curtain.hagl04"]:
+            if "proto" not in entity or int(entity["proto"][0:1]) == 1:
+                data_key = DATA_KEY_PROTO_V1
+            else:
+                data_key = DATA_KEY_PROTO_V2
+            entities.append(
+                XiaomiGenericCover(entity, "Curtain", data_key, gateway, config_entry)
```
```suggestion
    for device in gateway.devices["cover"]:
        model = device["model"]
        if model in ["curtain", "curtain.aq2", "curtain.hagl04"]:
            if "proto" not in device or int(device["proto"][0:1]) == 1:
                data_key = DATA_KEY_PROTO_V1
            else:
                data_key = DATA_KEY_PROTO_V2
            entities.append(
                XiaomiGenericCover(device, "Curtain", data_key, gateway, config_entry)
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/cover.py:
```diff
@@ -13,29 +14,31 @@
         self._data_key = data_key
         self._pos = 0
-        XiaomiDevice.__init__(self, device, name, xiaomi_hub)
+        XiaomiDevice.__init__(self, device, name, xiaomi_hub, config_entry)
```
```suggestion
        super().__init__(device, name, xiaomi_hub, config_entry)
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/light.py:
```diff
@@ -12,32 +12,35 @@
-    add_entities(devices)
+    entities = []
+    gateway = hass.data[DOMAIN][config_entry.entry_id]
+    for entity in gateway.devices["light"]:
+        model = entity["model"]
+        if model in ["gateway", "gateway.v3"]:
+            entities.append(
+                XiaomiGatewayLight(entity, "Gateway Light", gateway, config_entry)
```
```suggestion
    for device in gateway.devices["light"]:
        model = device["model"]
        if model in ["gateway", "gateway.v3"]:
            entities.append(
                XiaomiGatewayLight(device, "Gateway Light", gateway, config_entry)
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/light.py:
```diff
@@ -12,32 +12,35 @@
         self._brightness = 100
 
-        XiaomiDevice.__init__(self, device, name, xiaomi_hub)
+        XiaomiDevice.__init__(self, device, name, xiaomi_hub, config_entry)
```
```suggestion
        super().__init__(device, name, xiaomi_hub, config_entry)
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/lock.py:
```diff
@@ -20,27 +21,26 @@
-    async_add_entities(devices)
+    entities = []
+    gateway = hass.data[DOMAIN][config_entry.entry_id]
+    for entity in gateway.devices["lock"]:
+        model = entity["model"]
+        if model == "lock.aq1":
+            entities.append(XiaomiAqaraLock(entity, "Lock", gateway, config_entry))
```
```suggestion
    for device in gateway.devices["lock"]:
        model = device["model"]
        if model == "lock.aq1":
            entities.append(XiaomiAqaraLock(device, "Lock", gateway, config_entry))
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/sensor.py:
```diff
@@ -24,50 +25,70 @@
-                    XiaomiSensor(device, "Temperature", "temperature", gateway)
+    entities = []
+    gateway = hass.data[DOMAIN][config_entry.entry_id]
+    for entity in gateway.devices["sensor"]:
+        if entity["model"] == "sensor_ht":
+            entities.append(
+                XiaomiSensor(
+                    entity, "Temperature", "temperature", gateway, config_entry
```
```suggestion
    for device in gateway.devices["sensor"]:
        if device["model"] == "sensor_ht":
            entities.append(
                XiaomiSensor(
                    device, "Temperature", "temperature", gateway, config_entry
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/sensor.py:
```diff
@@ -24,50 +25,70 @@
-                devices.append(
-                    XiaomiSensor(device, "Temperature", "temperature", gateway)
+            )
+            entities.append(
+                XiaomiSensor(entity, "Humidity", "humidity", gateway, config_entry)
+            )
+        elif entity["model"] in ["weather", "weather.v1"]:
+            entities.append(
+                XiaomiSensor(
+                    entity, "Temperature", "temperature", gateway, config_entry
```
```suggestion
            entities.append(
                XiaomiSensor(device, "Humidity", "humidity", gateway, config_entry)
            )
        elif device["model"] in ["weather", "weather.v1"]:
            entities.append(
                XiaomiSensor(
                    device, "Temperature", "temperature", gateway, config_entry
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/sensor.py:
```diff
@@ -24,50 +25,70 @@
-                    XiaomiSensor(device, "Illumination", "illumination", gateway)
+            )
+            entities.append(
+                XiaomiSensor(entity, "Humidity", "humidity", gateway, config_entry)
+            )
+            entities.append(
+                XiaomiSensor(entity, "Pressure", "pressure", gateway, config_entry)
+            )
+        elif entity["model"] == "sensor_motion.aq2":
+            entities.append(
+                XiaomiSensor(entity, "Illumination", "lux", gateway, config_entry)
+            )
+        elif entity["model"] in ["gateway", "gateway.v3", "acpartner.v3"]:
+            entities.append(
+                XiaomiSensor(
+                    entity, "Illumination", "illumination", gateway, config_entry
```
```suggestion
                XiaomiSensor(device, "Humidity", "humidity", gateway, config_entry)
            )
            entities.append(
                XiaomiSensor(device, "Pressure", "pressure", gateway, config_entry)
            )
        elif device["model"] == "sensor_motion.aq2":
            entities.append(
                XiaomiSensor(device, "Illumination", "lux", gateway, config_entry)
            )
        elif device["model"] in ["gateway", "gateway.v3", "acpartner.v3"]:
            entities.append(
                XiaomiSensor(
                    device, "Illumination", "illumination", gateway, config_entry
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/sensor.py:
```diff
@@ -24,50 +25,70 @@
-                devices.append(
-                    XiaomiSensor(device, "Bed Activity", "bed_activity", gateway)
+            )
+        elif entity["model"] in ["vibration"]:
+            entities.append(
+                XiaomiSensor(
+                    entity, "Bed Activity", "bed_activity", gateway, config_entry
```
```suggestion
        elif device["model"] in ["vibration"]:
            entities.append(
                XiaomiSensor(
                    device, "Bed Activity", "bed_activity", gateway, config_entry
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/sensor.py:
```diff
@@ -24,50 +25,70 @@
-                devices.append(
-                    XiaomiSensor(device, "Tilt Angle", "final_tilt_angle", gateway)
+            )
+            entities.append(
+                XiaomiSensor(
+                    entity, "Tilt Angle", "final_tilt_angle", gateway, config_entry
```
```suggestion
            entities.append(
                XiaomiSensor(
                    device, "Tilt Angle", "final_tilt_angle", gateway, config_entry
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/sensor.py:
```diff
@@ -24,50 +25,70 @@
-                devices.append(
-                    XiaomiSensor(device, "Coordination", "coordination", gateway)
+            )
+            entities.append(
+                XiaomiSensor(
+                    entity, "Coordination", "coordination", gateway, config_entry
```
```suggestion
            entities.append(
                XiaomiSensor(
                    device, "Coordination", "coordination", gateway, config_entry
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/sensor.py:
```diff
@@ -24,50 +25,70 @@
-    add_entities(devices)
+            )
+        else:
+            _LOGGER.warning("Unmapped Device Model ")
```
```suggestion
            _LOGGER.warning("Unmapped Device Model")
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/sensor.py:
```diff
@@ -24,50 +25,70 @@
         """Initialize the XiaomiSensor."""
         self._data_key = data_key
-        XiaomiDevice.__init__(self, device, name, xiaomi_hub)
+        XiaomiDevice.__init__(self, device, name, xiaomi_hub, config_entry)
```
```suggestion
        super().__init__(device, name, xiaomi_hub, config_entry)
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/switch.py:
```diff
@@ -20,76 +21,108 @@
-                    XiaomiGenericSwitch(device, "Plug", data_key, True, gateway)
+    entities = []
+    gateway = hass.data[DOMAIN][config_entry.entry_id]
+    for entity in gateway.devices["switch"]:
+        model = entity["model"]
+        if model == "plug":
+            if "proto" not in entity or int(entity["proto"][0:1]) == 1:
+                data_key = "status"
+            else:
+                data_key = "channel_0"
+            entities.append(
+                XiaomiGenericSwitch(
+                    entity, "Plug", data_key, True, gateway, config_entry
```
```suggestion
    for device in gateway.devices["switch"]:
        model = device["model"]
        if model == "plug":
            if "proto" not in device or int(device["proto"][0:1]) == 1:
                data_key = "status"
            else:
                data_key = "channel_0"
            entities.append(
                XiaomiGenericSwitch(
                    device, "Plug", data_key, True, gateway, config_entry
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/switch.py:
```diff
@@ -20,76 +21,108 @@
-                        device, "Wall Switch", "channel_0", False, gateway
-                    )
+            )
+        elif model in ["ctrl_neutral1", "ctrl_neutral1.aq1"]:
+            entities.append(
+                XiaomiGenericSwitch(
+                    entity, "Wall Switch", "channel_0", False, gateway, config_entry
```
```suggestion
        elif model in ["ctrl_neutral1", "ctrl_neutral1.aq1"]:
            entities.append(
                XiaomiGenericSwitch(
                    device, "Wall Switch", "channel_0", False, gateway, config_entry
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/switch.py:
```diff
@@ -20,76 +21,108 @@
-                        device, "Wall Switch LN", "channel_0", False, gateway
-                    )
+            )
+        elif model in ["ctrl_ln1", "ctrl_ln1.aq1"]:
+            entities.append(
+                XiaomiGenericSwitch(
+                    entity, "Wall Switch LN", "channel_0", False, gateway, config_entry
```
```suggestion
        elif model in ["ctrl_ln1", "ctrl_ln1.aq1"]:
            entities.append(
                XiaomiGenericSwitch(
                    device, "Wall Switch LN", "channel_0", False, gateway, config_entry
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/switch.py:
```diff
@@ -20,76 +21,108 @@
+        elif model in ["ctrl_neutral2", "ctrl_neutral2.aq1"]:
+            entities.append(
+                XiaomiGenericSwitch(
+                    entity,
```
```suggestion
                    device,
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/switch.py:
```diff
@@ -20,76 +21,108 @@
+            )
+            entities.append(
+                XiaomiGenericSwitch(
+                    entity,
```
```suggestion
                    device,
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/switch.py:
```diff
@@ -20,76 +21,108 @@
+        elif model in ["ctrl_ln2", "ctrl_ln2.aq1"]:
+            entities.append(
+                XiaomiGenericSwitch(
+                    entity,
```
```suggestion
                    device,
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/switch.py:
```diff
@@ -20,76 +21,108 @@
+            )
+            entities.append(
+                XiaomiGenericSwitch(
+                    entity,
```
```suggestion
                    device,
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/switch.py:
```diff
@@ -20,76 +21,108 @@
-                    XiaomiGenericSwitch(device, "Wall Plug", data_key, True, gateway)
+            )
+        elif model in ["86plug", "ctrl_86plug", "ctrl_86plug.aq1"]:
+            if "proto" not in entity or int(entity["proto"][0:1]) == 1:
+                data_key = "status"
+            else:
+                data_key = "channel_0"
+            entities.append(
+                XiaomiGenericSwitch(
+                    entity, "Wall Plug", data_key, True, gateway, config_entry
```
```suggestion
            if "proto" not in device or int(device["proto"][0:1]) == 1:
                data_key = "status"
            else:
                data_key = "channel_0"
            entities.append(
                XiaomiGenericSwitch(
                    device, "Wall Plug", data_key, True, gateway, config_entry
```

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/switch.py:
```diff
@@ -20,76 +21,108 @@
         self._power_consumed = None
         self._supports_power_consumption = supports_power_consumption
-        XiaomiDevice.__init__(self, device, name, xiaomi_hub)
+        XiaomiDevice.__init__(self, device, name, xiaomi_hub, config_entry)
```
```suggestion
        super().__init__(device, name, xiaomi_hub, config_entry)
```

MartinHjelmare commented on tests/components/xiaomi_aqara/test_config_flow.py:
```diff
@@ -0,0 +1,324 @@
+TEST_PORT = 1234
+TEST_NAME = "Test_Aqara_Gateway"
+TEST_SID = "abcdefghijkl"
+TEST_PROTECOL = "1.1.1"
```
```suggestion
TEST_PROTOCOL = "1.1.1"
```

MartinHjelmare commented on tests/components/xiaomi_aqara/test_config_flow.py:
```diff
@@ -0,0 +1,324 @@
+        gateway.ip_adress = host
+        gateway.port = TEST_PORT
+        gateway.sid = TEST_SID
+        gateway.proto = TEST_PROTECOL
```
```suggestion
        gateway.proto = TEST_PROTOCOL
```

MartinHjelmare commented on tests/components/xiaomi_aqara/test_config_flow.py:
```diff
@@ -0,0 +1,324 @@
+        CONF_MAC: TEST_MAC,
+        const.CONF_INTERFACE: config_flow.DEFAULT_INTERFACE,
+        const.CONF_DISCOVERY_RETRY: config_flow.DEFAULT_DISCOVERY_RETRY,
+        const.CONF_PROTOCOL: TEST_PROTECOL,
```
```suggestion
        const.CONF_PROTOCOL: TEST_PROTOCOL,
```

MartinHjelmare commented on tests/components/xiaomi_aqara/test_config_flow.py:
```diff
@@ -0,0 +1,324 @@
+        CONF_MAC: TEST_MAC,
+        const.CONF_INTERFACE: config_flow.DEFAULT_INTERFACE,
+        const.CONF_DISCOVERY_RETRY: config_flow.DEFAULT_DISCOVERY_RETRY,
+        const.CONF_PROTOCOL: TEST_PROTECOL,
```
```suggestion
        const.CONF_PROTOCOL: TEST_PROTOCOL,
```

MartinHjelmare commented on tests/components/xiaomi_aqara/test_config_flow.py:
```diff
@@ -0,0 +1,324 @@
+        CONF_MAC: TEST_MAC,
+        const.CONF_INTERFACE: config_flow.DEFAULT_INTERFACE,
+        const.CONF_DISCOVERY_RETRY: config_flow.DEFAULT_DISCOVERY_RETRY,
+        const.CONF_PROTOCOL: TEST_PROTECOL,
```
```suggestion
        const.CONF_PROTOCOL: TEST_PROTOCOL,
```

MartinHjelmare commented on tests/components/xiaomi_aqara/test_config_flow.py:
```diff
@@ -0,0 +1,324 @@
+    assert result["step_id"] == "user"
+    assert result["errors"] == {}
+
+    result = await hass.config_entries.flow.async_configure(
```
We need to patch the calls to the library so no I/O is done in the tests.

starkillerOG replied:
I thought the pytest.fixture already took care of that (line 27-41)?:
```
@pytest.fixture(name="xiaomi_aqara", autouse=True)
def xiaomi_aqara_fixture():
    """Mock xiaomi_aqara discovery and entry setup."""
    mock_gateway_discovery = get_mock_discovery([TEST_HOST])

    with patch(
        "homeassistant.components.xiaomi_aqara.config_flow.XiaomiGatewayDiscovery",
        return_value=mock_gateway_discovery,
    ), patch(
        "homeassistant.components.xiaomi_aqara.config_flow.get_mac_address",
        return_value=TEST_MAC,
    ), patch(
        "homeassistant.components.xiaomi_aqara.async_setup_entry", return_value=True
    ):
        yield
```

MartinHjelmare replied:
Right. I missed that fixture. :+1:

MartinHjelmare committed:
Clean up

starkillerOG committed:
Update homeassistant/components/xiaomi_aqara/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
Update homeassistant/components/xiaomi_aqara/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
Update homeassistant/components/xiaomi_aqara/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
Update homeassistant/components/xiaomi_aqara/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
Update homeassistant/components/xiaomi_aqara/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
resolve data storage

starkillerOG commented on homeassistant/components/xiaomi_aqara/__init__.py:
```diff
@@ -217,21 +122,109 @@ def remove_device_service(call):
+    )
+    hass.data[DOMAIN][entry.entry_id] = xiaomi_gateway
+
+    if hass.data[DOMAIN].get("Listener") is None:
```
done

starkillerOG commented on homeassistant/components/xiaomi_aqara/__init__.py:
```diff
@@ -334,24 +363,26 @@ def parse_data(self, data, raw_data):
         sid = str(sid).replace(":", "").lower()
 
-        for gateway in xiaomi.gateways.values():
+        for gateway in hass.data[DOMAIN].values():
```
I fixed this by placing the gateways (multiple under their entry_id) under:
`hass.data[DOMAIN][GATEWAYS_KEY][entry.entry_id]`
and the listener (only one for all gateways) under:
`hass.data[DOMAIN][LISTENER_KEY]`

starkillerOG commented on homeassistant/components/xiaomi_aqara/config_flow.py:
```diff
@@ -0,0 +1,188 @@
+            self.interface = user_input[CONF_INTERFACE]
+
+            # Discover Xiaomi Aqara Gateways in the netwerk to get required SIDs.
+            xiaomi = XiaomiGatewayDiscovery(self.hass.add_job, [], self.interface)
```
That was already the case in the integration before I converted it to Config Flow.
This PR does not change anything to the underlying lib but only implements Config Flow.

So this is out of scope for this PR.
But indeed it is weird and a good side note....

starkillerOG committed:
move format_mac down

starkillerOG committed:
Remove discovery_retry from config flow

starkillerOG committed:
remove unused strings

starkillerOG committed:
fix styling

starkillerOG committed:
fix black styling

starkillerOG committed:
fix tests

starkillerOG committed:
remove mac connection

This is needed to prevent a conflict with the Xiaomi Miio integration that I discovered during testing.

starkillerOG committed:
fix flake8

starkillerOG commented:
@MartinHjelmare I have fully tested this.
I think it is now ready to be merged.
Only thing left is the unique_id (mac or sid), see the discussion about pros/cons...

starkillerOG commented:
I think the failed tests are unrelated to this PR, could you confirm @MartinHjelmare

starkillerOG committed:
remove getmac depandance

starkillerOG commented:
@MartinHjelmare any more feedback?

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/config_flow.py:
```diff
@@ -0,0 +1,167 @@
+
+            # Discover Xiaomi Aqara Gateways in the netwerk to get required SIDs.
+            xiaomi = XiaomiGatewayDiscovery(self.hass.add_job, [], self.interface)
+            await self.hass.async_add_executor_job(xiaomi.discover_gateways)
```
If the interface address is badly formatted or incorrect somehow there will probably be an error here that we should catch.

starkillerOG replied:
Nice, you are completly right!
I have added the exception, handeld the error and added a test for it.

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/config_flow.py:
```diff
@@ -0,0 +1,167 @@
+        if user_input is not None:
+            # get all required data
+            name = user_input[CONF_NAME]
+            key = user_input.get(CONF_KEY)
```
If the user enters an incorrect key, could that fail the connection to the gateway?

starkillerOG replied:
I don't think so, keys are only used to control subdevices, but is not nessesary to get status of sensors etc.

MartinHjelmare replied:
Still, wouldn't it be good to validate the key?

starkillerOG replied:
just tested it and I can confirm that if you enter an invalid key everything will work just fine, the component will be set up and all sensors will just work and get data.
Only if you try to control something, lets say turn on the gateway light, you will get this error:
`ERROR (SyncWorker_20) [xiaomi_gateway] Got error element in data {"error":"Invalid key"}`

starkillerOG replied:
@MartinHjelmare I don't think there is an easy way to test it.
If you want to test it you need to control some device, lets say toggle the gateway light. But I don't think it is nice to turn on/off the light while setting up an integration just to test...

But I could implement that if you want, simply turn on and then turn off the build in light of the gateway....

MartinHjelmare replied:
Using `write_to_hub` seems to validate the key.

https://github.com/Danielhiversen/PyXiaomiGateway/blob/0f7ba3b53d8c72260e602feb3bcd8079d435a0c9/xiaomi_gateway/__init__.py#L330

Can we send some dummy command that doesn't really do anything but still validates the key?

starkillerOG replied:
I implemented a key validation

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/config_flow.py:
```diff
@@ -0,0 +1,167 @@
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context.update({"title_placeholders": {"name": self.host}})
+
+        return await self.async_step_user()
```
Do we still need to ask the user for interface if we know the host address of the gateway?

starkillerOG replied:
yes because the interface is the address the socket binds to in order to receive messages from the gateway. that is not the gateway ip, but the ip of the device that is running HomeAssistant.
Or 'any', however 'any' does not always work (for me when I have VPN enabled on my dev PC 'any" does not work, withouth VPN it does work.

starkillerOG committed:
check for inavlid_interface + test

starkillerOG commented:
@MartinHjelmare can you now approve?

starkillerOG committed:
Validate gateway key

starkillerOG committed:
add invalid key tests

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/config_flow.py:
```diff
@@ -0,0 +1,183 @@
+                self.selected_gateway.key = key
+                valid_key = self.selected_gateway.write_to_hub(sid, mid=10000)
+            else:
+                valid_key = True
```
What is the use case for not providing a key?

starkillerOG replied:
The current integration also offers the capability of not providing a key.
I guess it is for people that can't get the key....
I know some gateway models are not able to retrieve the key (although I think with those gateways the whole port is closed so you also can't connect to the gateway for the sensors).

However I am not sure, why you would not specify a key...
I just did not want to break it for people that use the current integration without a key.

starkillerOG replied:
Without a key you can get the state of all devices and sensors, you just can't control any device. But If you only own sensors it doesn't really matter...

MartinHjelmare replied:
Should we then limit the platforms to only sensor platforms when there's no key set? The other platforms won't work without a key.

starkillerOG replied:
Good sugestion, I limited it to only sensors and binary_sensors if no key is provided

starkillerOG committed:
Fix spelling

starkillerOG committed:
Only set up sensors if no key

starkillerOG commented:
@MartinHjelmare tested the code again, and everything seems to work.
I think this is ready for merging.

MartinHjelmare reviewed (approved):
Good!

MartinHjelmare commented:
Please update the breaking change paragraph and mention that if no key is set only the binary_sensor and sensor platforms will be loaded. A key is required to activate the other platforms.

starkillerOG commented:
@MartinHjelmare I added it to the breaking change
